### ISSUE TITLE
[BUG] TypeError: Cannot read properties of undefined (reading 'info')

### ISSUE THREAD

**Issue by indrodev** [poolifier/poolifier#1468]
### Duplicate issue

- [x] I've searched for any related issues and avoided creating a duplicate issue.

### Description

TypeError: Cannot read properties of undefined (reading 'info')
    at exports.DynamicThreadPool.getWorkerInfo (/usr/src/app/node_modules/poolifier/lib/index.js:1:41573)
    at exports.DynamicThreadPool.shallUpdateTaskFunctionWorkerUsage (/usr/src/app/node_modules/poolifier/lib/index.js:1:34972)
    at exports.DynamicThreadPool.beforeTaskExecutionHook (/usr/src/app/node_modules/poolifier/lib/index.js:1:34178)
    at exports.DynamicThreadPool.executeTask (/usr/src/app/node_modules/poolifier/lib/index.js:1:42235)
    at /usr/src/app/node_modules/poolifier/lib/index.js:1:33317
    at new Promise (<anonymous>)
    at exports.DynamicThreadPool.execute (/usr/src/app/node_modules/poolifier/lib/index.js:1:32589)
    at HttpModeTask.process (/usr/src/app/dist/lib/Task/task.service.js:469:18)
    at HttpModeTask.execute (/usr/src/app/dist/lib/Task/task.service.js:288:32)
    at Object.processJob [as fn] (/usr/src/app/dist/utils/agenda/jobs/task.js:24:35) 

### Expected behavior

No errors

### Actual behavior

throws error

### Steps to reproduce

this error occurs after running for approximately 3 days continuously

### poolifier version

^2.7.4

### Node.js version

v18.18.0

### System

  System:
    OS: Linux 5.15 Zorin OS 16.3
    CPU: (6) x64 AMD Ryzen 5 3600 6-Core Processor
    Memory: 22.82 GB / 29.20 GB
    Container: Yes
    Shell: 5.8 - /usr/bin/zsh

**Comment by jerome-benoit** [comment]
It looks like race condition with worker node removal and querying. 

What is the code triggering it?

**Comment by indrodev** [comment]
FixedThreadPool working with no issues while testing, but when using DynamicThreadPool facing high RAM usage overtime > 1.5 GB to 2.5 GB (I might not implemented it correctly, but before using native node:worker_threads have not faced this issue before)

Please let me know if used this package incorrectly

### FixedThreadPool Example...

A simple fixed worker to fetch logs from docker containers, at a given interval (triggered via a cron-like service). Max time it should take to respond should be 59 secs (1000 * 59)

processModuleLogs.worker.ts

```typescript
import { ThreadWorker } from "poolifier";
import { z } from "zod";
import Docker, { ContainerInfo } from "dockerode";
import { DateTime } from "luxon";

const WorkerPayloadSchema = z.object({
    seedData: z.object({
        companyId: z.string(),
        deviceId: z.string(),
        deviceUserType: z.string(),
        integrationType: z.string(),
        label: z.string(),
        locationId: z.string(),
    }),
    duration: z.number(),
    excludeModules: z.array(z.string()),
});

export type WorkerPayload = z.infer<typeof WorkerPayloadSchema>;

export type LogsOptions = {
    follow: false;
    timestamps: true;
    stderr: true;
    stdout: true;
    since: number;
    until: number;
};

export type Stream = {
    stream: {
        label: string;
        deviceId: string;
        locationId: string;
        companyId: string;
        deviceUserType: string;
        integrationType: string;
        serviceName: string;
    };

    values: string[][];
};

export type WorkerResponse = {
    error: boolean;
    message: string;
    data: Stream[];
};

let docker: Docker;

function getTimeRange(duration: number) {
    const currentTime = DateTime.now().toUTC();
    const endTime = DateTime.fromJSDate(currentTime.toJSDate()).toSeconds();
    const startTime = DateTime.fromJSDate(currentTime.toJSDate())
        .minus({ minutes: duration })
        .toSeconds();

    return { startTime, endTime };
}

function formatLog(log: string[]) {
    if (!(Array.isArray(log) && log.length > 0)) {
        return [];
    }

    const resp: string[][] = [];
    log.forEach((el) => {
        if (el.length === 0) {
            return;
        }
        const data = el.split(" ");
        const regex = /\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z)/;
        const timestamp = data[0].match(regex);
        if (timestamp && timestamp.length > 0) {
            const nanoSec = String(DateTime.fromISO(timestamp[0]).toUTC().toMillis() * 1000000);
            data.shift();
            const packet = [nanoSec, data.join(" ")];
            resp.push(packet);
        }
    });
    return resp;
}

async function fetchContainerList(params: Pick<WorkerPayload, "excludeModules">) {
    const list = await docker.listContainers({ all: true });

    const filteredList = list.filter((el) => {
        const exists = el.Names.find((name) => name.includes("/service_"));
        if (!exists) {
            return false;
        }
        if (
            params.excludeModules &&
            params.excludeModules.length > 0 &&
            params.excludeModules.includes(exists.slice(1))
        ) {
            return false;
        }
        return true;
    });
    return filteredList;
}

function getRawModuleLog(id: string, options: LogsOptions): Promise<string> {
    return new Promise((resolve) => {
        const cont = docker.getContainer(id);

        cont.logs(options, (err, data) => {
            if (err) {
                console.error("unable to fetch logs for container id -> %s", id);
                return resolve("");
            }
            return resolve(data ? data.toString("utf8").trim() : "");
        });
    });
}

async function getProcessedModuleLog(
    module: ContainerInfo,
    options: LogsOptions,
    seedData: WorkerPayload["seedData"]
) {
    const name = module.Names.join("").slice(1);
    const containerId = module.Id;
    let log: string[][] | undefined = [];
    try {
        const rawLog = await getRawModuleLog(containerId, options);

        if (rawLog !== "") {
            log = formatLog(rawLog.split("\n"));
        }

        // process logs
    } catch (err) {
        console.warn("unable to find logs for %s [%s]", name, containerId);
    }

    return {
        stream: {
            ...seedData,
            serviceName: name,
        },
        values: log,
    } satisfies Stream;
}

async function fetchModulesLogs(options: WorkerPayload) {
    const { startTime, endTime } = getTimeRange(options.duration);
    const logsOptions = {
        follow: false,
        timestamps: true,
        stderr: true,
        stdout: true,
        since: startTime,
        until: endTime,
    } satisfies LogsOptions;

    const list = await fetchContainerList({ excludeModules: options.excludeModules });

    const logs = await Promise.allSettled(
        list.map((el) => getProcessedModuleLog(el, logsOptions, options.seedData))
    );

    const successLogs = logs.reduce((acc, el) => {
        if (el.status === "fulfilled" && el.value.values.length > 0) {
            acc.push(el.value);
        }
        return acc;
    }, [] as Stream[]);

    return {
        error: false,
        message: "ok",
        data: successLogs,
    };
}

class LogWorker extends ThreadWorker<WorkerPayload, Promise<WorkerResponse>> {
    constructor() {
        super(async (data) => await this.execute(data), {
            maxInactiveTime: 1000 * 59,
        });
    }

    private async execute(data: WorkerPayload | undefined): Promise<WorkerResponse> {
        try {
            const valid = await WorkerPayloadSchema.safeParseAsync(data);

            if (!valid.success) {
                return {
                    error: true,
                    message: "invalid worker payload",
                    data: [],
                };
            }

            docker = new Docker({
                socketPath: process.env.DOCKER_SOCKET_PATH,
            });

            const resp = await fetchModulesLogs({
                seedData: {
                    companyId: valid.data.seedData.companyId,
                    deviceId: valid.data.seedData.deviceId,
                    deviceUserType: valid.data.seedData.deviceUserType,
                    integrationType: valid.data.seedData.integrationType,
                    label: valid.data.seedData.label,
                    locationId: valid.data.seedData.locationId,
                },
                duration: valid.data.duration,
                excludeModules: valid.data.excludeModules,
            });

            return resp;
        } catch (err) {
            return {
                error: true,
                message: err instanceof Error ? err.message : "something went wrong in log worker",
                data: [],
            };
        }
    }
}

export default new LogWorker();

```

workerInterface.ts

```typescript
import path from "node:path";
import { WorkerPayload, WorkerResponse } from "../workers/log/processModuleLogs.worker.js";
import config from "../../config";
import { logger } from "../logger";
import { FixedThreadPool } from "poolifier";

const PROCESS_MODULE_LOG_WORKER_PATH = path.join(
    __dirname,
    "../workers/log",
    `/processModuleLogs.worker${path.extname(__filename)}`
);

export const poolGetModulesLogs = new FixedThreadPool<WorkerPayload, Promise<WorkerResponse>>(
    1,
    PROCESS_MODULE_LOG_WORKER_PATH,
    {
        workerOptions: {
            env: {
                DOCKER_SOCKET_PATH: config.DOCKER_SOCKET_PATH,
            },
        },
        errorHandler: (e: Error) => {
            logger.error("error in log worker %o", e);
            return e;
        },
        onlineHandler: () => {
            logger.info("Log Worker is online");
        },
    }
);
```

handleLog.js

```typescript
// ... some code ...
import * as LogWorkerInterface from "../../workerInterface/log";
// ... some code ...
const logs = await LogWorkerInterface.poolGetModulesLogs
                .execute({
                    seedData: {
                        companyId: system.data.companyId?.toString() ?? "NOT LINKED",
                        deviceId: system.data.deviceId ?? "NOT LINKED",
                        deviceUserType: "local",
                        integrationType: system.data.integrationType ?? "NOT LINKED",
                        label: "service",
                        locationId: system.data.locationId?.toString() ?? "NOT LINKED",
                    },
                    duration: 1,
                    excludeModules: config.EXCLUDE_LOG_MODULES,
                })
                .catch((err) => {
                    console.error(err);
                    return {
                        error: true,
                        message: "log processing error. look at schedule",
                        data: [],
                    };
                });

// ... some code ...
```


### DynamicThreadPool Example...

A simple file uploader (triggered via a cron-like service), each file size 350 - 500 KB (maximum 800 KB). Max time it should take to respond should be 5 mins (1000 * 60 * 5)

processUpload.worker.ts

```typescript
import { ThreadWorker } from "poolifier";
import { uploadToBlobStorage } from "../../uploadHelper";
import { IPayloadAttachment } from "../../mongodb/schema/Task.model";
import { IStorageDetails } from "../../mongodb/schema/System.model";

export type WorkerPayload = {
    payload: IPayloadAttachment[];
    storageList: IStorageDetails[];
};

export type WorkerResponse = IPayloadAttachment[];

export class ProcessUploadWorker extends ThreadWorker<WorkerPayload, Promise<WorkerResponse>> {
    constructor() {
        super(async (data) => await this.execute(data), {
            maxInactiveTime: 1000 * 60 * 5,
        });
    }

    private async execute(data: WorkerPayload | undefined): Promise<WorkerResponse> {
        if (!data) {
            throw new Error("invalid upload payload");
        }
        const resp = await uploadToBlobStorage(data.payload, data.storageList); // uploading file via azure sdk
        return resp;
    }
}

export default new ProcessUploadWorker();

```

workerInterface.ts

```typescript
import path from "node:path";
import { WorkerPayload, WorkerResponse } from "../workers/task/processHttp.worker";
import config from "../../config";
import { logger } from "../logger";
import { DynamicThreadPool, availableParallelism } from "poolifier";

const PROCESS_UPLOAD_WORKER_PATH = path.join(
    __dirname,
    "../workers/task",
    `/processUpload.worker${path.extname(__filename)}`
);

const MAX_POOL_SIZE = availableParallelism() > 1 ? availableParallelism() - 1 : 1;

export const poolProcessUpload = new DynamicThreadPool<WorkerPayload, Promise<WorkerResponse>>(
    1,
    MAX_POOL_SIZE,
    PROCESS_UPLOAD_WORKER_PATH,
    {
        errorHandler: (e: Error) => {
            logger.error("error in upload worker %o", e);
            return e;
        },
        onlineHandler: () => {
            logger.info("Upload Worker is online");
        },
    }
);
```

handleUpload.js

```typescript
// ... some code ...
import * as TaskWorkerInterface from "../../utils/workerInterface/task";
// ... some code ...
const attachments = await TaskWorkerInterface.poolProcessUpload
                .execute({
                    payload: this._task.payload.attachments as IPayloadAttachment[], // the files it will upload is in a b64 encoded string, and maximum items is 1
                    storageList: storageList,
                })
                .catch((err) => {
                    logger.error("%o", err);
                    return err;
                });

// ... some code ...
```

**Comment by jerome-benoit** [comment]
> FixedThreadPool working with no issues while testing, but when using DynamicThreadPool facing high RAM usage overtime > 1.5 GB to 2.5 GB (I might not implemented it correctly, but before using native node:worker_threads have not faced this issue before)

Looks like there's different issues: 
+ memory leak
+ null exception

The latter should be 'fixed' in 3.0.1.
 
> Please let me know if used this package incorrectly

Given the code surface, it will takes a bit of time ... 
If you can extract a self-contained repro case, that will make the memory leak debugging easier. 

Could you check that your code properly close ressources at worker destruction: socket, file handle, file stream, etc.? 
If not, please use the `killHandler`.

**Comment by indrodev** [comment]
Thanks for your response, will observe it for 1 week. If I find any more insights related to this issue, will post them here.
